docs/src/content/docs/book/assembly-functions.mdx

```diff
@@ -0,0 +1,678 @@
+asm fun bocchiThe(BOC: Cell): Cell { BOC }
+```
+
+The parameters of arbitrary [Struct][struct] types are distributed over their fields, recursively flattened as the arguments are pushed onto the stack. In particular, the value of the first field of the [Struct][struct] is pushed first, the second is pushed second, and so on, so that the value of the first field is at the bottom of the stack and the value of the last is at the top. If there are nested structures inside those [Structs][struct], they're flattened in the same manner.
+
+```tact
+// Struct with two fields of type Int
+struct AB { a: Int; b: Int }
+
+// This will produce the sum of two fields in the `AB` Struct
+asm fun sum(two: AB): Int { ADD }
+
+// Struct with two nested `AB` structs as its fields
+struct Nested { ab1: AB; ab2: AB }
+
+// This will multiply the sums of fields of nested `AB` Structs
+asm fun mulOfSums(n: Nested): Int { ADD -ROT ADD MUL }
+
+// Action!
+fun showcase() {
+    sum(AB{ a: 27, b: 50 }); // 77
+    //      ↑      ↑
+    //      |      Pushed last, sits on top of the stack
+    //      Pushed first, sits on the bottom of the stack
+
+    mulOfSums(Nested{ ab1: AB{ a: 1, b: 2 }, ab2: AB{ a: 3, b: 4 } }); // 21
+    //                         ↑     ↑                ↑     ↑
+    //                         |     |                |     Pushed last,
+    //                         |     |                |     sits on top of the stack
+    //                         |     |                Pushed second-to-last,
+    //                         |     |                sits below the top of the stack
+    //                         |     Pushed second,
+    //                         |     sits right above the bottom of the stack
+    //                         Pushed first, sits on the bottom of the stack
+}
+```
+
+### Returns {#conventions-returns}
+
+When present, return type of an assembly function attempts to capture relevant values from the resulting stack after the function execution and possible stack [arrangements](#arrangements). When not present, however, assembly function does not take any values from the stack.
+
+When present, an assembly function's return type attempts to grab relevant values from the resulting stack after the function execution and any [result arrangements](#arrangements). If the return type is not present, however, the assembly function does not take any values from the stack.
+
+```tact
+// Pushes `x` onto the stack, increments it there,
+// but does not capture the result, leaving it on the stack
+asm fun push(x: Int) { INC }
+```
+
+Specifying a [primitive type][p], such as an [`Int{:tact}`][int] or a [`Cell{:tact}`][cell], will make the assembly function pop the top value from the stack and produce it as a result. If the run-time type of the popped value doesn't match the specified return type, an exception with [exit code 7](/book/exit-codes#7) will be thrown: `Type check error`.
```

Comment:
confusing description: "pop the top value from the stack and produce it as a result"

afaiu, no popping happens, also unclear that "produce" means here

better say something like nothing happens to `x` below and it just sits on top of the stack and will be used as the return value

should we also make it clear that the asm function below by itself won't produce exit code 7, but only when it's called and the return value is used as a cell